concordium-node/src/grpc2.rs

```diff
@@ -2290,3 +2296,116 @@ impl<A> Require<tonic::Status> for Option<A> {
         }
     }
 }
+
+/// Tower layer adding middleware updating the node stats related to gRPC API.
+#[derive(Clone)]
+pub struct StatsLayer {
+    stats: Arc<StatsExportService>,
+}
+
+impl<S> tower::Layer<S> for StatsLayer {
+    type Service = StatsMiddleware<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        StatsMiddleware {
+            stats: self.stats.clone(),
+            inner: service,
+        }
+    }
+}
+
+/// Tower middleware for updating the node stats related to the gRPC API.
+#[derive(Clone)]
+pub struct StatsMiddleware<S> {
+    stats: Arc<StatsExportService>,
+    inner: S,
+}
+
+/// Get the label from a gRPC status code. This is used when updating metrics of
+/// the prometheus exporter.
+fn get_grpc_code_label(code: tonic::Code) -> &'static str {
+    use tonic::Code::*;
+    match code {
+        Ok => "ok",
+        Cancelled => "cancelled",
+        Unknown => "unknown",
+        InvalidArgument => "invalid argument",
+        DeadlineExceeded => "deadline exceeded",
+        NotFound => "not found",
+        AlreadyExists => "already exists",
+        PermissionDenied => "permission denied",
+        ResourceExhausted => "resource exhausted",
+        FailedPrecondition => "failed precondition",
+        Aborted => "aborted",
+        OutOfRange => "out of range",
+        Unimplemented => "unimplemented",
+        Internal => "internal",
+        Unavailable => "unavailable",
+        DataLoss => "data loss",
+        Unauthenticated => "unauthenticated",
+    }
+}
+
+/// Actual middleware implementation updating the stats.
+impl<S> tower::Service<hyper::Request<hyper::Body>> for StatsMiddleware<S>
+where
+    S: tower::Service<
+            hyper::Request<hyper::Body>,
+            Response = hyper::Response<tonic::body::BoxBody>,
+        > + Clone
+        + Send
+        + 'static,
+    S::Future: Send + 'static,
+{
+    type Error = S::Error;
+    type Future = futures::future::BoxFuture<'static, Result<Self::Response, Self::Error>>;
+    type Response = S::Response;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: hyper::Request<hyper::Body>) -> Self::Future {
+        // This is necessary because tonic internally uses `tower::buffer::Buffer`.
+        // See https://github.com/tower-rs/tower/issues/547#issuecomment-767629149
+        // for details on why this is necessary
+        let clone = self.inner.clone();
```

Comment:
Can you please link to the documentation https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services which explains this.

Reply:
I have updated the comment